File: src/main/java/land/leets/domain/portfolio/usecase/GetPortfoliosImpl.java
package land.leets.domain.portfolio.usecase;

import land.leets.domain.portfolio.domain.Portfolio;
import land.leets.domain.portfolio.domain.ProjectScope;
import land.leets.domain.portfolio.domain.repository.PortfolioRepository;
import land.leets.domain.portfolio.exception.PortfolioNotFoundException;
import land.leets.domain.portfolio.presentation.dto.PortfolioResponse;
import land.leets.domain.portfolio.presentation.dto.PortfoliosResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
public class GetPortfoliosImpl implements GetPortfolios {

    private final PortfolioRepository portfolioRepository;

    @Override
    public List<List<PortfoliosResponse>> all(String generation) {
        List<List<PortfoliosResponse>> response = new ArrayList<>();
        if (generation == null) {
            response.add(getPortfoliosByScope(ProjectScope.FINAL));
            response.add(getPortfoliosByScope(ProjectScope.TOY));
            return response;
        }
        response.add(getPortfoliosByGenerationAndScope(Long.parseLong(generation), ProjectScope.FINAL));
        response.add(getPortfoliosByGenerationAndScope(Long.parseLong(generation), ProjectScope.TOY));

        return response;
    }

    @Override
    public PortfolioResponse one(Long portfolioId) {
        Portfolio portfolio = portfolioRepository.findById(portfolioId)
                .orElseThrow(PortfolioNotFoundException::new);

        return PortfolioResponse.from(portfolio);
    }

    private List<PortfoliosResponse> getPortfoliosByScope(ProjectScope scope) {
        return portfolioRepository.findAllByScopeOrderByGenerationDesc(scope).stream()
                .map(PortfoliosResponse::from)
                .toList();
    }

    private List<PortfoliosResponse> getPortfoliosByGenerationAndScope(Long generation, ProjectScope scope) {
        return portfolioRepository.findAllByGenerationAndScope(generation, scope).stream()
                .map(PortfoliosResponse::from)
                .toList();
    }
}

File: src/main/kotlin/land/leets/domain/portfolio/usecase/GetPortfoliosImpl.kt
package land.leets.domain.portfolio.usecase

import land.leets.domain.portfolio.domain.repository.PortfolioRepository
import land.leets.domain.portfolio.exception.PortfolioNotFoundException
import land.leets.domain.portfolio.presentation.dto.PortfolioResponse
import land.leets.domain.portfolio.presentation.dto.PortfoliosResponse
import land.leets.domain.portfolio.type.ProjectScope
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class GetPortfoliosImpl(
    private val portfolioRepository: PortfolioRepository
) : GetPortfolios {

    @Transactional(readOnly = true)
    override fun all(generation: String?): List<List<PortfoliosResponse>> {
        if (generation == null) {
            return  listOf(
                getPortfoliosByScope(ProjectScope.FINAL),
                getPortfoliosByScope(ProjectScope.TOY)
            )
        }
        return listOf(
            getPortfoliosByGenerationAndScope(generation.toLong(), ProjectScope.FINAL),
            getPortfoliosByGenerationAndScope(generation.toLong(), ProjectScope.TOY)
        )
    }

    @Transactional(readOnly = true)
    override fun one(portfolioId: Long): PortfolioResponse {
        val portfolio = portfolioRepository.findById(portfolioId)
            .orElseThrow { PortfolioNotFoundException() }

        return PortfolioResponse.from(portfolio)
    }

    private fun getPortfoliosByScope(scope: ProjectScope): List<PortfoliosResponse> {
        return portfolioRepository.findAllByScopeOrderByGenerationDesc(scope)
            .map { PortfoliosResponse.from(it) }
            .toList()
    }

    private fun getPortfoliosByGenerationAndScope(
        generation: Long,
        scope: ProjectScope
    ): List<PortfoliosResponse> {
        return portfolioRepository.findAllByGenerationAndScope(generation, scope)
            .map { PortfoliosResponse.from(it) }
            .toList()
    }
}
